packages/ipfs/src/cli/commands/add.js

```diff
@@ -236,7 +243,10 @@ module.exports = {
         mode: argv.mode,
         mtime
       })
-      : getStdin() // Pipe directly to ipfs.add
+      : tagWithMetadata(getStdin(), {
+        mode: argv.mode,
+        mtime
+      }) // Pipe to ipfs.add tagging with mode and mtime
```

Comment:
Would this work, without having to use `normaliseAddInput`?:

```suggestion
      : {
        content: getStdin(),
        mode: argv.mode,
        mtime
      } // Pipe to ipfs.add tagging with mode and mtime
```

Reply:
Just gave it a go, but it seems `ipfs.add` is expecting an iterator:

> TypeError: Cannot read property 'Symbol(Symbol.asyncIterator)' of undefined

We are able to get the job done if we turn the object into an async iterator

```javascript
: (function * () {
   yield {
     content: getStdin(),
     mode: argv.mode,
     mtime
   }
 })()
```

Is there a iterator util for turning an element into a async iterator?

Reply:
Weird, that should be supported. You should also be able to do:

```javascript
: [{
     content: getStdin(),
     mode: argv.mode,
     mtime
   }]
```

Reply:
That worked. Should I open a bug for the add function when passed:

```javascript
{
   content: getStdin(),
   mode: argv.mode,
   mtime
}
```

Reply:
I think this is because the test stubs `ipfs.add` to only return something when passed an iterable (async or otherwise) as the first argument:

```javascript
ipfs.add.withArgs(matchIterable(), defaultAddArgs()).returns([{
  cid,
  path: 'readme'
}])
```

If you pass:

```javascript
{
   content: getStdin(),
   mode: argv.mode,
   mtime
}
```

You're not passing an iterable so the stub doesn't match and `undefined` is returned from `ipfs.add` because JavaScript which causes the reported `TypeError` in the `for await..of` loop.

The test needs to be updated to this:

```javascript
it('add from pipe with mtime=100', async () => {
  const cid = new CID('QmPZ9gcCEpqKTo6aq61g2nXGUhM4iCL3ewB6LDXZCtioEB')

  ipfs.add.withArgs(sinon.match({
    content: matchIterable(),
    mtime: { secs: 100 }
  }), defaultAddArgs()).returns([{
    cid,
    path: 'readme'
  }])

  const proc = cli('add --mtime=100', {
    ipfs,
    getStdin: function * () {
      yield Buffer.from('hello\n')
    }
  })

  const out = await proc
  expect(out).to.equal(`added ${cid} ${cid}\n`)

  const input = ipfs.add.getCall(0).args[0]
  expect(input).to.have.nested.property('mtime.secs', 100)
})
```

Reply:
TBH it could even lose the last assertion (for the `mtime.secs` property) as it'll only fail if the `ipfs.add` stub doesn't match, which will cause the `TypeError` issue before it gets to that assertion.

Reply:
@achingbrain I have updated the usage and test as suggested. The mocking had to be changed in 'add from pipe' as well.